Add OSX compressedrefs support in openj9-openjdk-jdk9

babsingh opened the pull request:
Set OPENJ9_BUILDSPEC to osx_x86-64_cmprssptrs for compressedrefs
build.

Append "-pagezero_size 0x1000" to LDFLAGS_JDKEXE on OSX 64-bit
compressedrefs build. This allows memory to be mapped below 4GB.

Ran ./common/autoconf/autogen.sh to re-generate the
generated-configure.sh scripts.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh commented:
Related to https://github.com/eclipse/openj9/issues/36.

fyi - @DanHeidinga

DanHeidinga reviewed (approved):
lgtm

DanHeidinga commented:
The copyright check is expected to fail for the generated files.

DanHeidinga commented:
@babsingh Is this ready?  I'm OK with merging once the WIP tag is removed,

babsingh commented:
Yes, these changes are ready. I wanted to verify if these changes work with the OpenJ9 spec file changes. Since these are front-end changes and other platforms use the same code, so these changes should work regardless. Thus, removing the WIP tag.

babsingh commented:
@DanHeidinga I have added `-pagezero_size` changes for `compressedrefs` build, and rebased to clear the conflicts. I can successfully build compressedrefs:

```
openjdk version "9.0.4-internal"
OpenJDK Runtime Environment (build 9.0.4-internal+0-adhoc.sbabneetcaibmcom.openj9-openjdk-jdk9)
Eclipse OpenJ9 VM (build osx_compressedrefs-e2a35e14a9, JRE 9 Darwin amd64-64-Bit Compressed References 20181010_000000 (JIT enabled, AOT enabled)
OpenJ9   - e2a35e14a9
OMR      - 57dd0f0b
JCL      - 47987161dc based on jdk-9.0.4+12)
```

DanHeidinga commented:
@keithc-ca Any objections to this approach?  I'm OK with it but not as well versed in the ojdk makefiles

keithc-ca commented on common/autoconf/flags.m4:
```diff
@@ -1222,7 +1222,7 @@ AC_DEFUN([FLAGS_SETUP_COMPILER_FLAGS_FOR_JDK_HELPER],
   # Customize LDFLAGS for executables
 
-  $2LDFLAGS_JDKEXE="${$2LDFLAGS_JDK}"
+  $2LDFLAGS_JDKEXE="${$2LDFLAGS_JDKEXE} ${$2LDFLAGS_JDK}"
```
Can this not be achieved without modifying openjdk source?

babsingh replied:
in `generated-configure.sh`, the closed OpenJ9 content is at the beginning, and then the common openjdk content is added. initial value specified in **closed/autoconf/custom-hook.m4** for `OPENJDK_BUILD_LDFLAGS_JDKEXE` and `LDFLAGS_JDKEXE` will be overwritten without the above change. not sure if there is way to achieve this without modifying openjdk source; couldn't find such an approach. but, we take a similar approach with `OPENJDK_BUILD_JAVA_BASE_LDFLAGS` and `JAVA_BASE_LDFLAGS`. unless the flags are predefined, the above modification to openjdk source should have no impact.

keithc-ca replied:
'spec.gmk' includes 'custom-spec.gmk', so the necessary adjustments can be made in `custom-spec.gmk.in`.

babsingh replied:
thanks @keithc-ca. i moved the changes into `custom-spec.gmk.in`.

keithc-ca commented on closed/autoconf/custom-spec.gmk.in:
```diff
@@ -76,3 +76,10 @@ ifeq ($(OPENJDK_TARGET_OS), windows)
+ifeq ($(OPENJDK_BUILD_OS), macosx)
+  ifeq ($(OPENJ9_LIBS_SUBDIR), compressedrefs)
+    # Set page zero size to 4KB for mapping memory below 4GB.
+    LDFLAGS_JDKEXE := @LDFLAGS_JDKEXE@ -pagezero_size 0x1000
```
This line could be more simply expressed as:
```
LDFLAGS_JDKEXE += -pagezero_size 0x1000
```

babsingh replied:
`+=` works. updated changes in PR.

babsingh committed:
Add OSX compressedrefs support

Set OPENJ9_BUILDSPEC to osx_x86-64_cmprssptrs for compressedrefs
build.

Append "-pagezero_size 0x1000" to LDFLAGS_JDKEXE on OSX 64-bit
compressedrefs build. This allows memory to be mapped below 4GB.

Ran ./common/autoconf/autogen.sh to re-generate the
generated-configure.sh scripts.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

DanHeidinga reviewed (approved):
lgtm
